src/olympia/blocklist/tasks.py

```diff
@@ -121,10 +146,81 @@ def upload_filter(generation_time, is_base=True):
             server.publish_record(stash_upload_data)
             statsd.incr('blocklist.tasks.upload_filter.upload_stash')
 
+    oldest_base_filter_id: int | None = None
+
+    # Get the oldest base_filter_id from the set of defined IDs
+    # We should delete stashes that are older than this time
+    for block_type in BlockType:
+        # Ignore soft blocked config timestamps if the switch is not active.
+        if block_type == BlockType.SOFT_BLOCKED and not waffle.switch_is_active(
+            'enable-soft-blocking'
+        ):
+            continue
+
+        if block_type in filter_list:
+            base_filter_id = generation_time
+        else:
+            base_filter_id = get_config(
+                # Currently we read from the old singular config key for
+                # hard blocks to preserve backward compatibility.
+                # In https://github.com/mozilla/addons/issues/15193
+                # we can remove this and start reading from the new plural key.
+                MLBF_BASE_ID_CONFIG_KEY(block_type, compat=True),
+                json_value=True,
+            )
+
+        if base_filter_id is not None:
+            if oldest_base_filter_id is None:
+                oldest_base_filter_id = base_filter_id
+            else:
+                oldest_base_filter_id = min(oldest_base_filter_id, base_filter_id)
```

Comment:
We could probably simplify this logic here. This diff currently passes all the tests:

```diff
diff --git a/src/olympia/blocklist/tasks.py b/src/olympia/blocklist/tasks.py
index 98b93d3ebd..70ff4e25ca 100644
--- a/src/olympia/blocklist/tasks.py
+++ b/src/olympia/blocklist/tasks.py
@@ -4,6 +4,7 @@ import re
 from datetime import datetime, timedelta
 from typing import List
 
+import waffle
 from django.conf import settings
 from django.contrib.admin.models import CHANGE, LogEntry
 from django.contrib.admin.options import get_content_type_for_model
@@ -114,6 +115,25 @@ def upload_filter(generation_time, filter_list=None, create_stash=False):
     old_records = server.records()
     attachment_types_to_delete = []
 
+    # We need to find the oldest base filter ID to clean up the Remote Settings
+    # collection. We start by collecting the base filter ID for each supported
+    # block type. These IDs might be updated when we upload a new filter.
+    base_filter_ids = dict()
+    for block_type in BlockType:
+        if block_type == BlockType.SOFT_BLOCKED and not waffle.switch_is_active(
+            'enable-soft-blocking'
+        ):
+            continue
+
+        base_filter_ids[block_type] = get_config(
+            # Currently we read from the old singular config key for hard
+            # blocks to preserve backward compatibility.
+            # In https://github.com/mozilla/addons/issues/15193, we can remove
+            # this and start reading from the new plural key.
+            MLBF_BASE_ID_CONFIG_KEY(block_type, compat=True),
+            json_value=True,
+        )
+
     if is_base:
         for block_type in filter_list:
             attachment_type = BLOCKLIST_RECORD_MLBF_BASE(block_type)
@@ -129,7 +149,9 @@ def upload_filter(generation_time, filter_list=None, create_stash=False):
                 # After we have succesfully uploaded the new filter
                 # we can safely delete others of that type
                 attachment_types_to_delete.append(attachment_type)
-
+            # Since a new base filter has been generated and uploaded, we
+            # update the filter ID in the dict of base filter IDs.
+            base_filter_ids[block_type] = generation_time
             statsd.incr('blocklist.tasks.upload_filter.upload_mlbf.base')
 
     # It is possible to upload a stash and a filter in the same task
@@ -145,25 +167,7 @@ def upload_filter(generation_time, filter_list=None, create_stash=False):
             server.publish_record(stash_upload_data)
             statsd.incr('blocklist.tasks.upload_filter.upload_stash')
 
-    oldest_base_filter_id: int | None = None
-
-    # Get the oldest base_filter_id from the set of defined IDs
-    # We should delete stashes that are older than this time
-    for block_type in BlockType:
-        base_filter_id = get_config(
-            # Currently we read from the old singular config key for
-            # hard blocks to preserve backward compatibility.
-            # In https://github.com/mozilla/addons/issues/15193
-            # we can remove this and start reading from the new plural key.
-            MLBF_BASE_ID_CONFIG_KEY(block_type, compat=True),
-            json_value=True,
-        )
-        if base_filter_id is not None:
-            if oldest_base_filter_id is None:
-                oldest_base_filter_id = base_filter_id
-            else:
-                oldest_base_filter_id = min(oldest_base_filter_id, base_filter_id)
-
+    oldest_base_filter_id = min(base_filter_ids.values())
     for record in old_records:
         # Delete attachment records that match the
         # attachment types of filters we just uplaoded
```

Reply:
I took this idea and updated tests to ensure we cover it properly. Then found a few issues, and made changes enough to fix the tests.